### ISSUE TITLE
How should in-place broadcasting with repeated indices behave?

### ISSUE THREAD

**Issue by mbauman** [JuliaLang/julia#31392]
In #31300, we noticed that there's a bit of a quirk with our current in-place broadcasting semantics:

```julia
julia> x = [1]
       x[[1,1]] .+= 1
       x
1-element Array{Int64,1}:
 2

julia> x = [1]
       @views x[[1,1]] .+= 1
       x
1-element Array{Int64,1}:
 3
```

Now, in general, we typically un-alias the values on the RHS from the destination to ensure we don't clobber their values, but we explicitly disable the unaliasing in cases where the LHS is `===` an argument on the RHS with the expectation that the iteration order is the same between the two arrays and thus won't go out-of-order with respect to eachother.  There's one case where that's not true — this case with repeated indices.  This puts in a strange place with respect to their semantics.

Ideally, I'd think that all four of these permutations should compute the same result:

```julia
julia> x = [3,4]
       @views x[[1,1]] .= x[[1,1]] .+ x[[1,1]]
       x
2-element Array{Int64,1}:
 6
 4

julia> x = [3,4]
       @views x[[1,1]] .+= x[[1,1]]
       x
2-element Array{Int64,1}:
 9
 4

julia> x = [3,4]
       v = view(x, [1,1])
       v .= v .+ v
       x
2-element Array{Int64,1}:
 12
  4

julia> x = [3,4]
       v = view(x, [1,1])
       v .+= v
       x
2-element Array{Int64,1}:
 12
  4
```

**Comment by bkamins** [comment]
CC @nalimilan (we will soon have the same issue in DataFrames.jl when we will be reimplementing `setindex!`)

**Comment by mbauman** [comment]
I'm a little stumped as to what the "right" answer to all my variations on `x[[1,1]] .+= x[[1,1]]` above should be... is that first element 6, 9, or 12?  Without worrying about how it's implemented, I'd say 9 is right out — it should either be view-y (and 12) or completely unaliased (and 6).  Given that we unalias other views in general, I'm leaning towards 6.

**Comment by stevengj** [comment]
We could just say that the result is undefined…

**Comment by mbauman** [comment]
I strongly dislike undefined behaviors and we unalias _everything_ else (without complaint!).  It took me a bit to see this clearly, but IMO this is just an overzealous optimization.

**Comment by chethega** [comment]
I think the ultimate issue is that a lot of machinery implicitly assumes that abstract arrays have the property that `a[i]` and `a[j]` are independent for `i != j`. E.g. the result of `a .= b` depends on iteration order. `sort!` does not sort.

I vote for documenting that abstract arrays are assumed to be internally unaliased, and views with repeating indices would be officially malformed input for all mutating base functions. I am happy with undefined results in that case.

```
julia> sort!(view(collect(1:10), rand(1:10, 10)))
10-element view(::Array{Int64,1}, [6, 8, 6, 10, 10, 7, 2, 9, 5, 7]) with eltype Int64:
  2
  8
  2
  5
  5
  6
 10
  7
  9
  6
```

**Comment by mbauman** [comment]
Yes, that's fair.  And my proposed fix is probably too conservative — a custom abstract array `A` will make a preventative copy in `A .+= 1` unless it extended the new `possibly_self_aliased` function.  That sucks.  That said, in the common case of a `SubArray` or `A[I] .+= ...` it's extraordinarily easy to check if any of the (outermost) indices are overlapping.  I think there's a middle-ground to be found here before totally throwing our hands into the air and copping out with a wholly undefined behavior.

**Comment by chethega** [comment]
But what I'm saying is that we can't fix this by copies. Suppose we have `A .= B`, both arrays are unaliased with each other, but `A` has internal aliasing. We can either specify iteration order for `A .= B`, or refuse to specify it. In the first case, we end up almost specifying the implementation, in the second case we declare our (perfectly fine) behavior as UB.

To me, the second case looks much cleaner: Multi-mutating methods (`broadcast!`, `map!`, `sort!`, multi-`setindex!`) assume that there is no internal aliasing of abstract arrays, and passing arguments with internal aliasing as the `dst` gives UB (typically something sensible). Otherwise, we would need to go through every single mutating function and document its behavior on internally aliased abstract arrays.

Similar to our pragmatic solution to boundschecks on views:
```
julia> A=view([-UInt(1)], [1]);
julia> A.indices[1][1]=-7; pointer(A.parent), A[1]
(Ptr{UInt64} @0x00007f0018a2b410, 0x00007f0018a2b410)
julia> A[1]=23; A
1-element view(::Array{UInt64,1}, [-7]) with eltype UInt64:

signal (11): Segmentation fault
```
The answer is "don't do that and test with --check-bounds=yes".

**Comment by tkf** [comment]
Instead of solving this at the level of broadcasting (#31407), can we return a read-only array from `view(x, [1, 1])`?  That is to say, `view` returns a read-only array unless it can prove that the indices are non-overlapping at the type-level.  This also "solves" the problem with `sort!` in the sense it throws a method error rather than returning a garbage.

Maybe we can have (say) `UniqueArray(x)` for asserting the elements in the array `x` are all unique and wrapping it into a custom array type.  We can also have `unsafe_uniquearray(x)` so that current `view(A, idx)` can be written as `view(A, unsafe_uniquearray(idx))`.
